velox/functions/prestosql/tests/ArrayShuffleTest.cpp

```diff
@@ -0,0 +1,310 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <numeric>
+#include <optional>
+#include <random>
+
+#include "velox/expression/VectorReaders.h"
+#include "velox/functions/prestosql/tests/utils/FunctionBaseTest.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::test;
+using namespace facebook::velox::functions::test;
+
+namespace {
+DecodedVector* decode(DecodedVector& decoder, const BaseVector& vector) {
+  SelectivityVector rows(vector.size());
+  decoder.decode(vector, rows);
+  return &decoder;
+}
+
+class ArrayShuffleTest : public FunctionBaseTest {
+ protected:
+  constexpr static const int32_t kNumOfEvalTimes = 100;
+  constexpr static const int32_t kNums = 10;
+  constexpr static const double kUniquenessRate = .7;
+
+  template <typename T>
+  void testExpr(
+      const std::vector<std::vector<std::optional<T>>>& expected,
+      const std::string& expression,
+      const std::vector<VectorPtr>& input) {
+    // 1. evaluate the given expression
+    auto arrayVectorResult =
+        evaluate<ArrayVector>(expression, makeRowVector(input));
+    // 2. decode element values from output ArrayVector
+    DecodedVector decoded;
+    exec::VectorReader<Any> reader(decode(decoded, *arrayVectorResult));
+
+    for (auto i = 0; i < arrayVectorResult->size(); i++) {
+      ASSERT_TRUE(reader.isSet(i))
+          << "Actual array at " << i
+          << " is a null, which we don't need to test here";
+      auto genericView = reader[i].castTo<Array<T>>();
+      std::vector<std::optional<T>> actualArray;
+      for (auto offset = 0; offset < genericView.size(); offset++) {
+        actualArray.push_back(genericView[offset]);
+      }
+      // 3. assert the two vectors containing the same elements (ignoring order)
+      ASSERT_TRUE(std::is_permutation(
+          actualArray.begin(), actualArray.end(), expected[i].begin()))
+          << "Actual array at " << i
+          << " must produce a permutation of the expected array";
+    }
+  }
+
+  /// To test a shuffle, we will do the following tests:
+  /// 1. run the test for T times with range-based input {0, 1, ..., n-1},
+  ///    and it should be at least X% times different in total.
+  ///    Let's say T = 100, n = 10, X = 70.
+  /// 2. test if each shuffle array is a permutation of the origin array,
+  ///    which ensures the shuffle will not miss any existing value or
+  ///    generate any unexpected values.
+  /// NOTE:
+  /// The combination of tests above is a naive but straightforward way of
+  /// verifying the shuffle's (1) randomness and (2) and correctness. However,
+  /// it doesn't guarantee the uniformity. We will avoid over testing with
+  /// leveraging standard deviation, etc.
+  void runTestKTimes(
+      const int32_t t = kNumOfEvalTimes,
+      const int32_t n = kNums) {
```

Comment:
Add another utility function to test if a shuffle array will be different from others most time.